kong/plugins/ldap-auth/access.lua

```diff
@@ -105,17 +106,20 @@ end
 
 
 local function cache_key(conf, username, password)
-  if not ldap_config_cache[conf] then
-    ldap_config_cache[conf] = md5(fmt("%s:%u:%s:%s:%u",
-                                      lower(conf.ldap_host),
-                                      conf.ldap_port,
-                                      conf.base_dn,
-                                      conf.attribute,
-                                      conf.cache_ttl))
+  local prefix = ldap_config_cache[conf]
+  if not prefix then
+    prefix = fmt("%s:%u:%s:%s:%u",
+                 lower(conf.ldap_host),
+                 conf.ldap_port,
+                 conf.base_dn,
+                 conf.attribute,
+                 conf.cache_ttl)
+
+    ldap_config_cache[conf] = prefix
   end
 
-  return fmt("ldap_auth_cache:%s:%s:%s", ldap_config_cache[conf],
-             username, password)
+  local suffix = fmt("%s:%s", prefix, username or "")
+  return fmt("ldap_auth_cache:%s", to_hex(hmac_sha1(password or "", suffix)))
```

Comment:
I think applying hmac_sha1 to everything (including the prefix) will make debugging problems in the cache more difficult. I suggest we only encript the username and password and leave the other values unencrypted. Something like this:

```
local function cache_key(conf, username, password)
  local prefix = ldap_config_cache[conf]
  if not prefix then
    prefix = fmt("ldap_auth_cache:%s:%u:%s:%s:%u:",
                 lower(conf.ldap_host),
                 conf.ldap_port,
                 conf.base_dn,
                 conf.attribute,
                 conf.cache_ttl)
    ldap_config_cache[conf] = prefix
  end

  return prefix .. to_hex(hmac_sha1(password or "", username or ""))
end
```

Note that I also moved `ldap_auth_cache` to the prefix there. This will reduce string concatenation and function calls at the expense of more space taken on the cache. I don't know which one is preferable.

Reply:
You do not need an hmac here. Obfuscating the password into a sha of the composite elements of the cache key should suffice. HMAC is used as an authentication mechanism to ensure a message hasn’t been tampered with in flight. 

If the hmac key isn’t a secret value, an hmac is the wrong construction.

Reply:
@kikito,

There is one reason to that:
makes `cache_key` always the same length and smaller, that has perfomance benefits

Reply:
> If the hmac key isn’t a secret value, an hmac is the wrong construction.

Yes, I know, just wanted to be collision free, but I can change it to sha1.

Reply:
@p0pr0ck5, the `hmac` is now changed to `sha1_bin`, I guess it is good enough, or should I go with `sha256`?

Reply:
I think sha1 is sufficient here 👍